File: day21.py
#!/usr/bin/env python3

import data21 as data

recipes = []
all_ingredients = {}
all_allergens = {}
for l in data.data.splitlines():
    ingredients_l, allergens_l = l.split('(contains ')
    ingredients = ingredients_l.split()
    allergens = allergens_l[:-1].split(', ')
    recipes.append((ingredients, allergens))
    for ingredient in ingredients:
        all_ingredients[ingredient] = True
    for allergen in allergens:
        all_allergens[allergen] = True
all_ingredients = all_ingredients.keys()
all_allergens = all_allergens.keys()
#print('recipes', recipes)
#print('all_ingredients', all_ingredients)
#print('all_allergens', all_allergens)

ingredient_allergen_counts = {}
for ingredients, allergens in recipes:
    for ingredient in ingredients:
        for allergen in allergens:
            ingredient_allergen_counts[ingredient] = ingredient_allergen_counts.get(ingredient, {})
            ingredient_allergen_counts[ingredient][allergen] = ingredient_allergen_counts[ingredient].get(allergen, 0)
            ingredient_allergen_counts[ingredient][allergen] += 1
#print('ingredient_allergen_counts', ingredient_allergen_counts)

recipe_allergen_counts = {}
for recipe, allergens in recipes:
    for allergen in allergens:
        recipe_allergen_counts[allergen] = recipe_allergen_counts.get(allergen, 0)
        recipe_allergen_counts[allergen] += 1
#print('recipe_allergen_counts', recipe_allergen_counts)

ingredients_no_allergens = []
for ingredient in all_ingredients:
    possible_allergens = 0
    for allergen, count in ingredient_allergen_counts[ingredient].items():
        if count == recipe_allergen_counts[allergen]:
            possible_allergens += 1
    if not possible_allergens:
        ingredients_no_allergens.append(ingredient)
#print('ingredients_no_allergens', ingredients_no_allergens)

appearances = 0
for ingredients, allergens in recipes:
    for ingredient in ingredients:
        if ingredient in ingredients_no_allergens:
            appearances += 1
#print('appearances', appearances)
print(appearances)

File: day21.hs
import qualified Data.Either as E
import qualified Data.List as L
import qualified Data.Map as M
import qualified Data.Set as S
import qualified Text.ParserCombinators.Parsec as PCPS

eol = PCPS.char '\n'

-- fhmbh jlsqx ... (contains eggs, wheat, sesame)
inputLine = do
    ingredients <- PCPS.many1 $ (PCPS.many1 PCPS.alphaNum) <* (PCPS.string " ")
    PCPS.string "(contains "
    allergens <- PCPS.sepBy (PCPS.many1 PCPS.alphaNum) (PCPS.string ", ")
    PCPS.string ")"
    return (S.fromList ingredients, S.fromList allergens)

inputFile =
    PCPS.many1 (inputLine <* eol) <* PCPS.eof

allergenToIngredientSetOfRecipe :: (S.Set String, S.Set String) -> [(String, S.Set String)]
allergenToIngredientSetOfRecipe (ingredients, allergens) =
    [(allergen, ingredients) | allergen <- S.elems allergens]

allergenToIngredientSetOfRecipes :: [(S.Set String, S.Set String)] -> [(String, S.Set String)]
allergenToIngredientSetOfRecipes recipes =
    L.concatMap allergenToIngredientSetOfRecipe recipes

allergenToIngredientMapOfRecipes :: [(S.Set String, S.Set String)] -> M.Map String (S.Set String)
allergenToIngredientMapOfRecipes recipes =
    M.fromListWith S.intersection $ allergenToIngredientSetOfRecipes recipes

allIngredientsOfRecipes :: [(S.Set String, S.Set String)] -> S.Set String
allIngredientsOfRecipes recipes =
    S.unions $ map fst recipes

ingredientMapIngredients :: M.Map String (S.Set String) -> S.Set String
ingredientMapIngredients ingredientMap =
    S.unions $ M.elems ingredientMap

ingredientMapNonIngredients :: S.Set String -> M.Map String (S.Set String) -> S.Set String
ingredientMapNonIngredients allIngredients ingredientMap =
    S.difference allIngredients $ ingredientMapIngredients ingredientMap

countIngredients :: [(S.Set String, S.Set String)] -> M.Map String Int
countIngredients recipes =
    M.fromListWith (+) $ L.concatMap ingredientTupleOnes recipes
    where
        ingredientTupleOnes (ingredients, allergens) = map (\i -> (i, 1)) $ S.elems ingredients

answer :: [(S.Set String, S.Set String)] -> Int
answer recipes = appearances
    where
        allIngredients = allIngredientsOfRecipes recipes
        ingredientMap = allergenToIngredientMapOfRecipes recipes
        nonAllergenIngredients = ingredientMapNonIngredients allIngredients ingredientMap
        counts = countIngredients recipes
        nonAllergenCounts = M.restrictKeys counts nonAllergenIngredients
        appearances = sum $ M.elems nonAllergenCounts

main :: IO ()
main = do
    let fn = "day21.txt"
    inputText <- readFile fn
    let input = E.fromRight [] $ PCPS.parse inputFile fn inputText
    putStrLn $ show $ answer $ input
